experimental/sql/dialects/rel_tree.py

```diff
@@ -0,0 +1,223 @@
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from ast import Param
+from dataclasses import dataclass
+from typing import Any, List
+from xdsl.ir import Block, Region, Operation, SSAValue, ParametrizedAttribute, Data, MLContext, Attribute
+from xdsl.dialects.builtin import StringAttr, ArrayAttr, ArrayOfConstraint, IntegerAttr
+from xdsl.irdl import AttributeDef, OperandDef, ResultDef, RegionDef, SingleBlockRegionDef, irdl_attr_definition, irdl_op_definition, ParameterDef, AnyAttr, VarOperandDef, builder
+
+# This file contains the relational tree dialect. This dialect represents
+# relational queries in a tree form, simplifying certain optimizations.
+
+
+@irdl_attr_definition
+class DataType(ParametrizedAttribute):
+  """
+  Models a datatype in a relational query.
+  """
+  name = "rel_tree.datatype"
+
+
+@irdl_attr_definition
+class Int32(DataType):
+  """
+  Models a int32 type in a relational query.
+
+  Example:
+
+  ```
+  !rel_tree.int32
+  ```
+  """
+  name = "rel_tree.int32"
+
+
+@irdl_attr_definition
+class String(DataType):
+  """
+  Models a string type in a relational query, that can be either nullable or
+  not.
+
+  Example:
+
+  ```
+  !rel_tree.string<0: !i1>
+  ```
+  """
+  name = "rel_tree.string"
+
+  nullable = ParameterDef(IntegerAttr)
+
+  @staticmethod
+  @builder
+  def get(val: int) -> 'String':
+    return String([IntegerAttr.from_int_and_width(val, 1)])
+
+
+@irdl_op_definition
+class Column(Operation):
+  """
+  References a specific column with name `col_name`.
+
+  Example:
+
+  ```
+  rel_tree.column() ["col_name" = "a"]
+  ```
+  """
+  name = "rel_tree.column"
+
+  col_name = AttributeDef(StringAttr)
+
+  @builder
+  @staticmethod
+  def get(table: Region, col_name: str) -> 'Column':
+    return Column.build(attributes={"col_name": StringAttr.from_str(col_name)},
+                        regions=[table])
+
+
+@irdl_op_definition
+class Predicate(Operation):
```

Comment:
The semantics of this operation rather seem to suggest the name `Comparison`, no? For me, a "predicate" is any expression that returns a Boolean, i.e., Boolean operators such as `And` and `Or` are also predicates.

Reply:
I agree and I renamed it to `compare`, since it feels more like an action to me.